---
title: "Overlap with chip-seq"
author: "wancen"
date: "2/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
## Load required libraries
suppressMessages(suppressWarnings(library(data.table)))
suppressMessages(suppressWarnings(library(plyranges)))
suppressMessages(suppressWarnings(library(tidyverse)))
suppressMessages(suppressWarnings(library(devtools)))
# install_github("Wancen/airpart")
library(airpart)
```


```{r}
mESC<-read_narrowpeaks("../data/mESC.narrowPeak",genome_info = "mm9")
colnames(mcols(mESC))<-c("name","score","fold_enrichment","-log(pvalue)","-log(FDR)/log(10)","summit")
mESC_filter<-mESC[which(mcols(mESC)$fold_enrichment>10)]

mef<-read_narrowpeaks("../data/mef.narrowPeak",genome_info = "mm9")
colnames(mcols(mef))<-c("name","score","fold_enrichment","-log(pvalue)","-log(FDR)/log(10)","summit")
mef_filter<-mef[which(mcols(mef)$fold_enrichment>10)]

suppressMessages(suppressWarnings(library(GenomicFeatures)))
suppressMessages(suppressWarnings(library("org.Mm.eg.db")))
txdb <- makeTxDbFromGFF("ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M1/gencode.vM1.annotation.gtf.gz")
g <- genes(txdb)
g$sym <- mapIds(org.Mm.eg.db, sub("\\..*","",names(g)), "SYMBOL", "ENSEMBL")
g <- g[!is.na(g$sym)]
# define promoter regions
gene_prom <- promoters(x = g, upstream = 50000, downstream = 50000)



## Get allelic imbalance genes
get_overlap<-function(gene_prom,feat,mef,mESC){
  poi<-which(mcols(gene_prom)$sym %in% feat)
  g2<-gene_prom[poi]
  
  mESC_ov<-join_overlap_inner(g2, mESC)
  mef_ov<-join_overlap_inner(g2, mef)
  
 
  mESC_ov2 <- mESC_ov %>%
    group_by(sym) %>%
    summarize(
      fold_enrichment=max(fold_enrichment),
      peak_count=sum(!is.na(fold_enrichment)))
  
  mef_ov2 <- mef_ov %>%
    group_by(sym) %>%
    summarize(
      fold_enrichment=max(fold_enrichment),
      peak_count=sum(!is.na(fold_enrichment)))
  
  enrich<-merge(mESC_ov2,mef_ov2,by="sym",all=T)
  return(enrich)
}

```

```{r}
load("../data/larsson_postprocess.rda")
nct<-length(cell_meta_unique) # number of cell types
ratio<-c57_fil/total_fil
ratio_pseudo<-(c57_fil+2)/(total_fil+4)
cluster<-genecluster(ratio_pseudo,nct,G=c(4,12,20,24,28,32))
table(cluster)
```
# Genes of interest
```{r}
i=7
feat<-which(cluster==i) # choose genes location
cl_ratio<-as.vector(unlist(ratio[feat,]))
cl_total<-as.vector(unlist(total_fil[feat,]))
dat<-tibble(ratio=cl_ratio,x=factor(rep(as.vector(celltypeQC[,1]),each=length(feat)),levels = cell_meta_unique),cts=cl_total)
summary<-dat %>% group_by(x) %>% summarise(weighted.mean=weighted.mean(ratio,cts,na.rm = T),mean=mean(ratio,na.rm = T),var=var(ratio,na.rm = T)) %>% data.frame()
summary2<-summary[order(summary$weighted.mean),]
# if(max(diff(summary2$weighted.mean))>0.1){
    print(paste("cluster",i,"has",length(feat),"genes exists allelic ratio imbalance"))
    print(knitr::kable(summary)) #check summary table
    overlaptxb<-get_overlap(gene_prom,names(feat),mef_filter,mESC_filter)
    print(knitr::kable(overlaptxb))
# }
# overlap<-as.data.frame(overlaptxb)
# overlap[is.na(overlap)] <- 0
# wilcox.test(overlap$peak_count.x,overlap$peak_count.y,alternative = "g",paired = T) # 0 replace NA
# wilcox.test(overlaptxb$peak_count.x,overlaptxb$peak_count.y,alternative = "g",paired = T) # still NA
int_genes<-c(sum(overlaptxb$peak_count.x,na.rm = T),sum(overlaptxb$peak_count.y,na.rm = T))

```

# Other genes
```{r}
i=1
feat2<-which(cluster==i) # choose genes location
cl_ratio<-as.vector(unlist(ratio[feat2,]))
cl_total<-as.vector(unlist(total_fil[feat2,]))
dat<-tibble(ratio=cl_ratio,x=factor(rep(as.vector(celltypeQC[,1]),each=length(feat2)),levels = cell_meta_unique),cts=cl_total)
summary<-dat %>% group_by(x) %>% summarise(weighted.mean=weighted.mean(ratio,cts,na.rm = T),mean=mean(ratio,na.rm = T),var=var(ratio,na.rm = T)) %>% data.frame()
print(knitr::kable(summary)) #check summary table
summary2<-summary[order(summary$weighted.mean),]
print(paste("cluster",i,"has",length(feat2),"genes exists allelic ratio imbalance"))

library(pbapply)
othergenes<-replicate(100,sample(feat2,length(feat),replace = TRUE)) #sample 100 times
genename<-row.names(ratio_pseudo)
othergene_name<-pbsapply(1:100,function(i) genename[othergenes[,i]])
test<-pbsapply(1:100,function(i) get_overlap(gene_prom, othergene_name[,i], mef_filter,mESC_filter))
# overlaptxb2<-get_overlap(gene_prom,othergenes[,2],mef_filter,mESC_filter)
# print(knitr::kable(overlaptxb2))


other_genes<-pbsapply(1:100, function(i) return(c(sum(test[[i]]$peak_count.x,na.rm = T),sum(test[[i]]$peak_count.y,na.rm = T))))


# overlap2<-as.data.frame(overlaptxb2)
# overlap2[is.na(overlap2)] <- 0
# wilcox.test(overlap2$peak_count.x,overlap2$peak_count.y,paired = T) # 0 replace NA
# wilcox.test(overlaptxb2$peak_count.x,overlaptxb2$peak_count.y,paired = T) # still NA

```
# chi-square test
```{r}
obj<-pbsapply(1:100,function(i){
  res<-chisq.test(matrix(c(int_genes,other_genes[,i]),ncol=2,byrow = T))
  return(res$p.value)
})

```
