---
title: "Deng2014"
author: "wancen"
- name: Wancen Mu, Hirak Sarkar, Avi Srivastava, Kwangbom Choi, Rob Patro, Michael I. Love
date: "January 13, 2021"
abstract: |
  airpart is an R package that identifies subsets of genes displaying differential CTS allelic imbalance across all cell types.

  It consists of two paragraphs.
tags: [fusedlasso, Single cell]
output: 
  html_document:
    toc: true  
    toc_float: true 
    theme: united  
    highlight: tango  
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(airpart)
```

# Load data

First, we load the preprocessed [Deng 2014](https://www.nature.com/articles/s41586-018-0836-1) data. It includes one maternal allelic counts(`b6`) and paternal allelic counts(`cast`) which was derived after removing some low quality cells indicated in the previous paper. `celltype` is a character vector of all cell types.

```{r}
suppressPackageStartupMessages(library(SingleCellExperiment))
load("../data/Deng.rda")
cell_meta_unique <- c("zy","early2cell","mid2cell","late2cell","4cell","8cell","16cell","earlyblast","midblast","lateblast") 
order_of_development <- order(match(celltype,cell_meta_unique))
assay.list<-list(a1=b6[,order_of_development],a2=cast[,order_of_development])
coldata<-data.frame(x=factor(celltype[order_of_development],levels=cell_meta_unique))
sce<-SingleCellExperiment(assays=assay.list, colData=coldata)
```

# Create Ratio matrix

Here we add the pseudo count for gene clustering and visualization.
```{r}
sce <- preprocess(sce)
```

Take a look at highest expressed genes
```{r}
counts(sce)[order(rowSums(counts(sce)),decreasing = T)[1:10],sample(dim(sce)[2],5)]
```
# Quality control step
## QC on cells
```{r}
cellQCmetrics<-cellQC(sce,mad_detected=4,mad_spikegenes = 5,spike = "Ercc")
cellQCmetrics
```

Now define cell filtering automatically or users can manually filter out based on `sum`,`detected` and `spikePercent`. 
```{r}
keep_cell <- (
  cellQCmetrics$filter_sum & # sufficient features (genes)
  cellQCmetrics$filter_detected & # sufficient molecules counted
  cellQCmetrics$filter_spike  # sufficient features expressed compared to spike genes, high quality cells
) 
table(keep_cell)
sce<-sce[,keep_cell]
```

## QC on genes
Note that we require genes expressed in at least **25%** cells within each cell types. 
```{r}
featureQCmetric<-featureQC(sce,spike = "Ercc",sd = 0.03)
keep_feature<-(featureQCmetric$filter_celltype &
                 featureQCmetric$filter_sd &
                 featureQCmetric$filter_spike)
table(keep_feature)
sce<-sce[keep_feature,]
```



```{r eval=FALSE, echo=FALSE}
# library(DESeq2)
# library(pbapply)
# sce2 <- SingleCellExperiment(list(counts=counts(sce)))
# check <- pbsapply(levels(sce$x), function(c) {
#   poi <- which(sce$x == c)
#   a<-rowSums(counts(sce[,poi]))
#   o <- order(a, decreasing=TRUE)
#   hist(unlist(counts(sce)[o[1:100],poi]),breaks=100)
#   dds <- DESeqDataSet(sce2[o[1:200],poi], design=~1)
#   dds <- estimateSizeFactors(dds)
#   dds <- estimateDispersions(dds)
#   plotDispEsts(dds)
# })
```
# Gene clustering

We provide two methods for gene clustering. 


## Exclude chromosome X genes 

Sex typeing allelic imbalance are excluded because we want to focus on genetically-driven AI.
And `GMM` can actually successfully detected chr x and some imprinted genes in one gene cluster as below figure shows there are monoallelic pattern among fibroblast cells but in a coordinate way. So after removing all the genes on chromosome X, we redo the gene cluster step.

```{r}
load("../data/mm9_genome.rda")
gene_chrx<- unname(g$sym[which(seqnames(g)=="chrX")])
sce <- sce[-which(rownames(sce)%in%gene_chrx),]
```

## * Gaussian Mixture modeling

```{r}
sce<-geneCluster(sce, G=c(8))
metadata(sce)$geneCluster
```

## * Hierarchical clustering 
```{r,eval=FALSE}
sce.hc<-geneCluster(sce,method="hierarchical")
metadata(sce.hc)$geneCluster
```

## Simple summary table of alleic ratio

Quickly look at the weighted mean of allelic ratio for each gene cluster. Identify the interested gene cluster. Note that we recommend users only run `airpart` when the largest ordered allelic ratio difference > 0.05 for speed concerns. We find that the allelic ratio of most of the gene clusters in such cases (small absolute allelic ratio differences) won't provide enough evidence to detect differential allelic imbalance.
Here we find out all clusters exists differential alleic imbalance. 
```{r}
summary<-summaryAllelicRatio(sce)
sapply(1:length(summary), function(i) {
  inst <- summary[[i]]
  inst_order <- inst[order(inst$weighted.mean), ]
  max(diff(inst_order$weighted.mean)) > 0.05
})
```
Here we use gene cluster 8 as an example. Since it has higher dispersion, we will use nonparametric approach to partitioning the cell types.

# Estimate overdispersion parameter

```{r}
estDisp(sce,genecluster = 8)
sce_sub <- fusedLasso(sce,
  model = "binomial",
  genecluster = 8, ncores = 4
)
knitr::kable(metadata(sce_sub)$partition,row.names = F)
```


Plotting a heatmap to show the cell types partition:

```{r}
jpeg(file="C:/Users/wancenmu/OneDrive - University of North Carolina at Chapel Hill/Lab/Github/airpartpaper/plot/heatmap_Deng.jpeg",width = 8, height = 7,units = "in",res=450)
makeHeatmap(sce_sub,order_by_group = FALSE,show_row_names = T)
dev.off()
```
```{r}
sce_sub <- allelicRatio(sce_sub)
```

```{r}
sce <- sce_sub
part <- metadata(sce)$part
part$coef <- 1/(1+exp(-part$coef))
part$x <- factor(part$x,levels = cell_meta_unique)
p <- ggplot2::ggplot(part, aes(x = .data$x, y = .data$coef)) +
  ggplot2::geom_point(aes(color = .data$part)) +
  ggplot2::geom_step(group = 1) +
  ggplot2::geom_hline(yintercept = 0.5, colour = "gray40", linetype = "dashed") +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    panel.grid.minor = element_blank(),
    panel.border = element_rect(fill = "transparent"),
    axis.text.x = element_text(angle = 90, vjust = 1, hjust=1,size=8)
  ) +
  labs(x = "cell types", y = "allelic ratio")
p
```

## GO term analysis

Through GO term evaluation, we can find indeed `regulation of myeloid cell differentiation` and differential `gene expression` GO terms has significant p value.
```{r, results="asis"}
library(goseq)
i=8
feat<-which(rowData(sce)$cluster==i)
genes <- rep(0,nrow(sce))
names(genes) <- rownames(sce)
genes[feat] <- 1
table(genes)
suppressMessages(pwf <- nullp(genes, "mm9", "geneSymbol"))
suppressMessages(gores <- goseq(pwf, "mm9", "geneSymbol")) 
p_adjust <- p.adjust(gores$over_represented_pvalue,method = "fdr")
gores$`p.adjust`<-p_adjust
knitr::kable(head(gores[gores$ontology=="BP" & gores$numInCat>10,], 15))
```


# Session Info
```{r}
sessionInfo()
```
